File: Tag/Dynamic-Programming/Easy/70.爬楼梯.java
/*
 * @lc app=leetcode.cn id=70 lang=java
 *
 * [70] 爬楼梯
 */

// @lc code=start
class Solution {
    public int climbStairs(int n) {
        int p = 0, q = 0, r = 1;
        for (int i = 1; i <= n; i ++) {
            p = q;
            q = r;
            r = p + q;
        }
        return r;
    }
}
// @lc code=end

// 复杂度分析
// 时间复杂度：循环执行 nnn 次，每次花费常数的时间代价，故渐进时间复杂度为 O(n)O(n)O(n)。
//空间复杂度：这里只用了常数个变量作为辅助空间，故渐进空间复杂度为 O(1)O(1)O(1)。

File: jsTest/dynamic-programming/easy/70.爬楼梯.js
/*
 * @lc app=leetcode.cn id=70 lang=javascript
 *
 * [70] 爬楼梯
 * 假设你正在爬楼梯。需要 n 阶你才能到达楼顶。
 * 每次你可以爬 1 或 2 个台阶。你有多少种不同的方法可以爬到楼顶呢？
 * 
 * 1 <= n <= 45
 * 
 * https://leetcode.cn/problems/climbing-stairs/solutions/286022/pa-lou-ti-by-leetcode-solution/?source=vscode
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var climbStairs = function(n) {
    let p = 0, q = 0, r = 1;
    for (let i = 1; i <= n; ++i) {
        p = q;
        q = r;
        r = p + q;
    }
    return r;
};

// let result = climbStairs(66)
// console.log(result)
// @lc code=end

